tests/integration/p2p.go

```diff
@@ -451,7 +309,61 @@ func getAllP2PCollections(
 	assert.Equal(s.t, expectedCollections, cols)
 }
 
-// waitForSync waits for all given wait channels to receive an item signaling completion.
+// waitForUpdateEvents waits for all selected nodes to publish an update event.
+//
+// Expected document heads will be updated for all nodes that should receive network merges.
+func waitForUpdateEvents(
+	s *state,
+	nodeID immutable.Option[int],
+	collectionID int,
+) {
+	if len(s.nodeUpdateSubs) == 0 {
+		return // skip network testing
+	}
+
+	ctx, cancel := context.WithTimeout(s.ctx, subscriptionTimeout*10)
+	defer cancel()
+
+	for i := 0; i < len(s.nodes); i++ {
+		if nodeID.HasValue() && nodeID.Value() != i {
+			continue // node is not selected
+		}
+
+		var evt event.Update
+		select {
+		case msg := <-s.nodeUpdateSubs[i].Message():
+			evt = msg.Data.(event.Update)
+		case <-ctx.Done():
+			s.t.Fatalf("timeout waiting for update event")
+		}
+
+		// update the actual document heads
+		s.actualDocHeads[i][evt.DocID] = evt.Cid
+
+		// update the expected document heads of connected nodes
+		for id := range s.nodeConnections[i] {
+			if _, ok := s.actualDocHeads[id][evt.DocID]; ok {
+				s.expectedDocHeads[id][evt.DocID] = evt.Cid
+			}
+		}
+		// update the expected document heads of replicator sources
+		for id := range s.nodeReplicatorTargets[i] {
+			if _, ok := s.actualDocHeads[id][evt.DocID]; ok {
+				s.expectedDocHeads[id][evt.DocID] = evt.Cid
+			}
+		}
```

Comment:
thougth: It's odd that we would be updating the expected heads only if already present in the actual heads. Shouldn't it be updated regardless?

Reply:
I believe the code is correct based on what we discussed regarding replicator target nodes sending updates of a document back to the replicator source node.